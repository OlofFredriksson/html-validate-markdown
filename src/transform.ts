import { Source, Transformer } from "html-validate";
import { TransformContext } from "html-validate/dist/transform";

function findLocation(
    source: string,
    index: number,
    preamble: number
): [number, number] {
    let line = 1;
    let prev = 0;
    let pos = source.indexOf("\n");
    while (pos !== -1) {
        if (pos > index) {
            return [line, index - prev + preamble + 1];
        }
        line++;
        prev = pos;
        pos = source.indexOf("\n", pos + 1);
    }
    return [line, 1];
}

function* markdownTransform(
    this: TransformContext,
    source: Source
): Iterable<Source> {
    const codeFence = /^(```([^\s]+))([^]*?)^```/gm;

    let match;
    while ((match = codeFence.exec(source.data)) !== null) {
        const [, preamble, lang, data] = match;
        const [line, column] = findLocation(
            source.data,
            match.index,
            preamble.length
        );
        const cur: Source = {
            data,
            offset: match.index + (source.offset || 0) + preamble.length,
            filename: source.filename,
            line,
            column,
            originalData: source.originalData || source.data,
        };

        /* unless the language is explicitly html the language is tested if it
         * have a configured transformer */
        const chain = `${source.filename}:${lang}`;
        if (lang === "html" || this.hasChain(chain)) {
            yield* this.chain(cur, chain);
        }
    }
}

markdownTransform.api = 1;

export default markdownTransform as Transformer;
