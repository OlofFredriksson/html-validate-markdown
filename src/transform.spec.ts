import transform from "./transform";

test("should extract html blocks from markdown files", () => {
    const result = transform("./test/markdown.md");
    expect(result).toHaveLength(2);

    expect(result[0].data).toMatchSnapshot();
    expect(result[0].filename).toEqual("./test/markdown.md");
    expect(result[0].line).toEqual(3);
    expect(result[0].column).toEqual(9);

    expect(result[1].data).toMatchSnapshot();
    expect(result[1].filename).toEqual("./test/markdown.md");
    expect(result[1].line).toEqual(9);
    expect(result[1].column).toEqual(9);
});

test("should extract html blocks from markdown files with multi line html", () => {
    const result = transform("./test/multiline.md");
    expect(result).toHaveLength(1);
});
