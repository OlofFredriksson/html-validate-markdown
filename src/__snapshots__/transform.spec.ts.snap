// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should extract html blocks from markdown files 1`] = `
"
<button type=\\"submit\\">Button</button>
"
`;

exports[`should extract html blocks from markdown files 2`] = `
"
<a href=\\"http://\\" title=\\"link\\">link</a>
"
`;
